parsec/scheduling.c

```diff
@@ -175,7 +175,9 @@ int __parsec_execute( parsec_execution_stream_t* es,
                              chore_id);
 #endif
         parsec_hook_t *hook = tc->incarnations[chore_id].hook;
-        assert( NULL != hook );
+        if( NULL == hook ) {
+            goto next_chore; /* Gracefuly manage the case when NO chores are available for this task */
```

Comment:
I think this is coming up regularly, but as far as I recall the code it is not correct. The chores are packed, no NULL chores can exists except at the end. Second, this loop is only executed for valid chores (aka not NULL hook). So, the only way to get into this is if no eval function accepted the work, in which case the correct outcome is to bail out of the loop (not go to the next chore).

Reply:
Yes I remembered this discussion and checked the history https://github.com/ICLDisco/parsec/pull/642#discussion_r1522166794 Summary of last time: I dropped it because I couldn't reproduce it anymore in the dtd case I was looking at.

Now it is back. What happens here is that we have only 1 valid chore (the JDF declares only a GPU chore), but we have runtime disabled it (set hook=NULL by force) because we didn't find a device of that type during the task pool creation. Not sure there is anything related to evaluate, looking into more details.

Reply:
ok, that makes sense. But, at the end, it is an incorrect application, as we do not have support for all tasks available. If we stop this at a single process, it will be difficult to properly report and act on this. Thus, I think this might be something that needs to be checked upfront, before the first task is even executed, and all processes must act uniformly.

Reply:
The chore is NULL because it is a dyld function, and when we tried to populate the dyld functions,  the device was not active, so we didn't search for it at all. The proposed behavior in that case seems reasonable to me. Discuss.

Reply:
The device was not active when we tried to dlsym but it is now ? If now the device is active, that's bad, the devices should have been freezed before. If the device is still inactive then this is an incorrect application and we should have detected it before. The problem is that at this point there is no easy way to bail out, we have to abort the entire app.

Reply:
The app is correct (it has only cuda incarnations, but that is not a bug), but we have detected (late) that the app cannot run on this system (no cuda devices, before and after freeze). 

Ideally we should not call `parsec_fatal` when we capture HOOK_RETURN_ERROR (whatever the reason), but instead bubble up the error condition all the way to `taskpool_wait` to return as an rc code, but that's work for another day. When we do that, the error cases discovered here will be handled correctly. 

Last thing left is to fix the CI to use the correct tag. Thomas wanted to wait for Geri's rework of the CI, but it may be pretty easy to do now.

Reply:
As I said in my previous message, we cannot escalate this across a distributed run, so from user perspective it might not be the best solution. The user should have tested upfront if it can start a GPU-only run on the setup.

Reply:
this all went away with the preselection logic in #648